File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    cnt = 0
    stack = []
    for i in range(len(number)):
        if len(stack) == 0:
            stack.append(number[i])
        else:
            while stack:
                if stack[-1]<number[i]:
                    stack.pop()
                    cnt += 1
                    if cnt == k:
                        return ''.join(stack) + number[i:]
                else:
                    break
            stack.append(number[i])
            if cnt == k:
                return ''.join(stack)+number[i+1:]
    return ''.join(stack[0:len(number)-k])

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;
class Solution {
    public String solution(String number, int k) {
        int start = 0;
        int tmp = 0;
        Stack<Integer> st = new Stack<>();
        Stack<Integer> st_input = new Stack<>();
        StringBuilder sb = new StringBuilder();
        for(int i = 0;i<number.length();i++){
            tmp = Integer.parseInt(String.valueOf(number.charAt(i)));
            while(!st.isEmpty() && k>0 && tmp>st.peek()){
                st.pop();
                k--;
            }
            st.push(tmp);
        }
        while (k > 0 && !st.isEmpty()) {
            st.pop();
            k--;
        }
        while(!st.isEmpty()){
            st_input.push(st.pop());
        }
        while(!st_input.isEmpty()){
            sb.append(st_input.pop());
        }
        return sb.toString();
    }
}